### PR TITLE
feat: df-2-doe windows

### PR DESCRIPTION
**Pull request by TfedUD**

### CODE REVIEW

**Comment by TfedUD** [comment]
ah crap I didn't reset the branch correct and all the ole commits are in there

**Comment by TfedUD** [comment]
hmm idk why that's failing its passing on my machine

**Comment by mostaphaRoudsari** [comment]
@TfedUD, I reset the branch to master and pushed your changes. Make sure to pull the changes to your local before pushing any new changes. Do NOT try to merge the changes in your branch. There is no changes. Just rest yours.

File: dragonfly_doe2/doe/floor_space.py [hunk 1]
```diff
@@ -92,6 +93,65 @@ def __repr__(self):
...
+    name: str
+    location: int
+    x: float
+    y: float
+    width: float
+    height: float
+
+    @classmethod
+    def from_window(cls, nm, x, y, w, h):
+        return cls(name=nm, x=x, y=y, width=w, height=h)
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/floor_space.py:113]: You don't need a `classmethod` for this one. This is basically the same as the default `__init__`.

File: dragonfly_doe2/doe/floor_space.py [hunk 2]
```diff
@@ -92,6 +93,65 @@ def __repr__(self):
...
+@dataclass
+class Window:
+    """*.inp Window object.
+
+    Init methods(s):
+        1. from_window(name, x, y, width, height)
+
+    """
+    name: str
+    location: int
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/floor_space.py:105]: What is location? I don't see it being set in the class method.

File: dragonfly_doe2/doe/floor_space.py [hunk 3]
```diff
@@ -92,6 +93,65 @@ def __repr__(self):
...
+
+    def __repr__(self):
+        return self.to_inp()
+
+
+@dataclass
+class WindowSet:
+    name: str
+    location: int
+    window_data: List[Window]
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/floor_space.py:131]: what about windows for a list of windows?

```suggestion
    windows: List[Window]
```

File: dragonfly_doe2/doe/floor_space.py [hunk 4]
```diff
@@ -92,6 +93,65 @@ def __repr__(self):
...
+
+    @classmethod
+    def from_params(cls, name: str, location: int,  rw_paras: RectWind):
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/floor_space.py:134]: Some docstring will be helpful here. I'm still not sure what is `location` supposed to be.

File: dragonfly_doe2/doe/floor_space.py [hunk 5]
```diff
@@ -92,6 +93,65 @@ def __repr__(self):
...
+        origins = [(orig.x, orig.y) for orig in rw_paras.origins]
+        widths = rw_paras.widths
+        heights = rw_paras.heights
+
+        windows = []
+        for i, (org, w, h) in enumerate(zip(origins, widths, heights)):
+            windows.append(
+                Window(
+                    f'{name}_wndw_{i}', location+1, org[0],
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/floor_space.py:143]:
```suggestion
                    f'{name}_wndw_{i}', location + 1, org[0],
```

File: dragonfly_doe2/doe/floor_space.py [hunk 6]
```diff
@@ -3,6 +3,7 @@
 from dataclasses import dataclass
 from typing import List
 from .utils import short_name
+from dragonfly.windowparameter import RectangularWindows as RectWind
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/floor_space.py:6]: Why do we need to shorten the name here. Removing Windows made me really confused as RectWind sound like a single Window to me.

File: dragonfly_doe2/doe/floor_space.py [hunk 7]
```diff
@@ -92,6 +93,65 @@ def __repr__(self):
...
+        heights = rw_paras.heights
+
+        windows = []
+        for i, (org, w, h) in enumerate(zip(origins, widths, heights)):
+            windows.append(
+                Window(
+                    f'{name}_wndw_{i}', location+1, org[0],
+                    org[1],
+                    w, h))
+        return cls(name=name, location=location+1, window_data=windows)
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/floor_space.py:146]:
```suggestion
        return cls(name=name, location=location + 1, window_data=windows)
```

File: dragonfly_doe2/doe/floor_space.py [hunk 8]
```diff
@@ -177,15 +237,37 @@ def from_room(cls, room: Room2D):
...
+        wall_objs = [
             Wall.from_room_seg(name, i, wall_constr_name)
-            for i, bc in enumerate(room.boundary_conditions)
+            for i, bc in enumerate(bcs)
             if str(bc) == 'Outdoors'
         ]
+
+        walls = []
+        for i, (wall, window) in enumerate(zip(wall_objs, windows)):
+            walls.append('\n'.join([wall.to_inp(), window.to_inp()]))
```
Comment (mostaphaRoudsari) [inline dragonfly_doe2/doe/floor_space.py:269]: If this is why you wanted to make the implementation object oriented I'm in favor of doing it now. Basically you want to create the windows when creating the walls. That is pretty straight forward and you should only move this code under the Wall class.

**Review by mostaphaRoudsari — CHANGES_REQUESTED** [review]
Hi @TfedUD, See my comments. I think you should make some changes in the implementation before we merge this PR in. That said, this is very close!

File: dragonfly_doe2/doe/floor_space.py (hunk 8 above)
Comment (TfedUD) [reply dragonfly_doe2/doe/floor_space.py:269]: oooo for sure! I'll definitely absorb this into the wall class! that makes a ton of sense and feels more 'right'!

Comment (TfedUD) [reply dragonfly_doe2/doe/floor_space.py:269]: @ Navy tomorrow so will be able to knock this out while the sun is still coming up!

**Comment by TfedUD** [comment]
@mostaphaRoudsari did I do the 'pull first' right?
It's not the 'pull with rebase' correct?

File: dragonfly_doe2/doe/floor_space.py (hunk 1 above)
Comment (TfedUD) [reply dragonfly_doe2/doe/floor_space.py:113]: sounds good!  this felt a bit verbose; whole-heartedly concur, the choice was a 50/50 "continuity/uniformity vs common sense"  and the Navy makes the option on the left the instinct 😅

File: dragonfly_doe2/doe/floor_space.py (hunk 2 above)
Comment (TfedUD) [reply dragonfly_doe2/doe/floor_space.py:105]:
```python
class WindowSet:
    name: str
    location: int
    window_data: List[Window]

    @classmethod
    def from_params(cls, name: str, location: int,  rw_paras: RectWind):
        origins = [(orig.x, orig.y) for orig in rw_paras.origins]
        widths = rw_paras.widths
        heights = rw_paras.heights

        windows = []
        for i, (org, w, h) in enumerate(zip(origins, widths, heights)):
            windows.append(
                Window(
                    f'{name}_wndw_{i}', location+1, org[0], ######  Arg 2
                    org[1],
                    w, h))
        return cls(name=name, location=location+1, window_data=windows)

    def to_inp(self):
        return '\n'.join([wndw.to_inp() for wndw in self.window_data])

    def __repr__(self):
        return self.to_inp()
```
have location in there to be a part of the 'serialization' of the window objs names to mitigate duplicate name issues

**Comment by mostaphaRoudsari** [comment]
Hi @TfedUD, you should have used `git reset` to reset to the latest commit. It is again messed up right now. I can try to fix the git history tomorrow morning and then review the changes.

**Comment by TfedUD** [comment]
@mostaphaRoudsari  DAMNIT lol *bigfacepalm*

**Comment by mostaphaRoudsari** [comment]
I cleaned it up! Also cleaned up the other new branch on top of this one.
image

**Review by mostaphaRoudsari — APPROVED** [review]
LGTM! There are parts that we can clean up later.

I'll merge this in so we can move forward quickly and fix the export for the apertures.

**Pull request merged**